src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+    elif new.properties is not None:
+        merge_lists(existing.properties, new.properties)
+
+
+def generate_component_list(packages) -> List[CycloneDxComponent]:
+    ref_component_map = {}
+    components = []
+    for package in packages:
+        hashes = get_hashes_list(package)
+        refs = get_external_refs(package)
+        licenses = get_licenses(package)
+        author = get_author_from_parties(package.get("parties"))
+        purl = package.get("purl")
+
+        name = package.get("name")
+        version = package.get("version")
+        #if we don't have at least the required name and version we skip the component
+        if name is not None and version is not None:
+            component = CycloneDxComponent(
+                name=package.get("name"), version=package.get("version"),
+                group=package.get("namespace"), purl=purl,
+                author=author, copyright=package.get("copyright"),
+                description=package.get("description"),
+                hashes=hashes, licenses=licenses, externalReferences=refs,
+                bom_ref=purl)
+
+            if purl not in ref_component_map.keys():
+                components.append(component)
+                ref_component_map[purl] = component
+            else:
+                merge_components(ref_component_map[purl], component)
+
+    return components
+
+
+def build_bom(codebase) -> CycloneDxBom:
+    # TODO: find out if we can always expect that header to be present
+    scancode_header = codebase.get_headers()[0]
+
+    scancode_version = scancode_header.get("tool_version")
+
+    tool_header = get_tool_header(scancode_version)
+    bom_metadata = CycloneDxMetadata(tools=[tool_header])
+
+    files = OutputPlugin.get_files(codebase)
+
+    # get all packages that are not None
+    packages = [package for file in files for package in file.get("packages")]
+    # associate dependency relationship by purl of dependent package
+    dep_map = dict([(package.get("purl"), package.get("dependencies"))
+                    for package in packages])
+
+    components = generate_component_list(packages)
+    # associate components by purl
+    comp_map = dict(map(lambda c: (c.purl, c), components))
+
+    dependencies = generate_dependencies_list(dep_map, comp_map)
+
+    bom = CycloneDxBom(components=components, metadata=bom_metadata, dependencies=dependencies)
+    return bom
+
+
+def truncate_none_or_empty_values(obj) -> dict:
+    """gets a dict from an object and drops all items
+     that have keys which are either None, an empty list or an empty dict"""
+    predicate = lambda el: not (el is None or
+                                (isinstance(el, list) or isinstance(el, dict))
+                                and len(el) == 0)
+    if hasattr(obj, '__dict__'):
+        obj = vars(obj)
+    obj_dict = {k: v for k, v in obj.items() if predicate(v)}
+    return obj_dict
+
+
+class CycloneDxEncoder(json.JSONEncoder):
```

Comment:
Could we instead have an `to_dict()` methods on objects to serialize and avoid having JSON-specific serialization code?

Reply:
> I am open to this but would like to understand your reasoning about it first. Is there some downside to the serialization code that I am not seeing?

We also use a `to_dict()` function throughout the codebase as a convention. 
A JSONEncoder makes the serialization JSON-specific and impossible to use with other presentations such as YAML and I would rather have an explicit field-by-field mapping rather than something implicit. 
Also why would you rename fields?  Does the format demands to have "bom-ref" vs. "bom_ref" ?